### DISCUSSION TITLE
contructed simplified invoice fails upon fatoora validation.

### DISCUSSION THREAD

**Discussion by asifmeer** [mrsool/zatca#34] [Q&A]
getting this errors: fatoora -validate -invoice signed_invoice.xml 
       
********** Welcome to ZATCA E-Invoice Java SDK 3.3.3 *********************
This SDK uses Java to call the SDK (jar) passing it an invoice XML file.
It can take a Standard or Simplified XML, Credit Note, or Debit Note.
It returns if the validation is successful or shows errors where the XML validation fails.
It checks for syntax and content as well.
You can use the command (fatoora -help) for more information.

****************************************************************
2024-07-10 16:05:39,798 [ERROR] XsdValidator - Schema validation failed due to SAXParseException
2024-07-10 16:05:39,804 [INFO] ValidationProcessorImpl - [XSD] validation result : FAILED
2024-07-10 16:05:39,804 [ERROR] ValidationProcessorImpl - xsd validation errors : 
2024-07-10 16:05:39,804 [ERROR] ValidationProcessorImpl - CODE : SAXParseException, MESSAGE : Schema validation failed; XML does not comply with UBL 2.1 standards in line with ZATCA specifications. ERROR 1: org.xml.sax.SAXParseException; lineNumber: 2; columnNumber: 7; cvc-elt.1.a: Cannot find the declaration of element 'root'.
2024-07-10 16:05:40,701 [INFO] ValidationProcessorImpl - [EN] validation result : PASSED
2024-07-10 16:05:40,702 [ERROR] ValidationProcessorImpl - en validation warnings : 
2024-07-10 16:05:40,702 [WARN] ValidationProcessorImpl - CODE : BR-CO-13, MESSAGE : [BR-CO-13]-Invoice total amount without VAT (BT-109) = Σ Invoice line net amount (BT-131) - Sum of allowances on document level (BT-107) + Sum of charges on document level (BT-108).
2024-07-10 16:05:41,052 [INFO] ValidationProcessorImpl - [KSA] validation result : PASSED
2024-07-10 16:05:41,100 [ERROR] QrGenerationService - failed to generate qr [unable to get seller name from invoice]
2024-07-10 16:05:41,101 [INFO] ValidationProcessorImpl - [QR] validation result : FAILED
2024-07-10 16:05:41,101 [ERROR] ValidationProcessorImpl - qr validation errors : 
2024-07-10 16:05:41,101 [ERROR] ValidationProcessorImpl - CODE : QRCODE_INVALID, MESSAGE : Invalid QR code format, Please follow the ZATCA QR code specifications
2024-07-10 16:05:41,558 [ERROR] SignatureValidator - Error : null
2024-07-10 16:05:41,558 [INFO] ValidationProcessorImpl - [SIGNATURE] validation result : FAILED
2024-07-10 16:05:41,558 [ERROR] ValidationProcessorImpl - signature validation errors : 
2024-07-10 16:05:41,558 [ERROR] ValidationProcessorImpl - CODE : invoiceSignedDataDigestValue, MESSAGE : wrong invoice hashing
2024-07-10 16:05:41,558 [ERROR] ValidationProcessorImpl - CODE : NullPointerException, MESSAGE : null
2024-07-10 16:05:41,558 [ERROR] ValidationProcessorImpl - CODE : signatureValue, MESSAGE : wrong signature Value 
2024-07-10 16:05:41,563 [INFO] ValidationProcessorImpl - [PIH] validation result : FAILED
2024-07-10 16:05:41,563 [ERROR] ValidationProcessorImpl - pih validation errors : 
2024-07-10 16:05:41,563 [ERROR] ValidationProcessorImpl - CODE : KSA-13, MESSAGE : PIH is inValid
2024-07-10 16:05:41,563 [INFO] InvoiceValidationService -  *** GLOBAL VALIDATION RESULT = FAILED

**Comment by asifmeer** [comment]
i am using sandbox sdk certificates.

here is my test_construct_invoice.rb code

require 'zatca'

puts "zatca construct invoice started"
#===============================================================================
# 1. Setup the Invoice Values
#===============================================================================
invoice_id = "SME00010"
invoice_uuid = "8e6000cf-1a98-4174-b3e7-b5d5954bc10d"
note = "ABC"
note_language_id = "ar"
issue_date = "2022-08-17"
issue_time = "17:41:08"

invoice_subtype = ZATCA::UBL::InvoiceSubtypeBuilder.build(
  simplified: true,
  third_party: false,
  nominal: false,
  exports: false,
  summary: false,
  self_billed: false
) # => "0200000"

payment_means_code = ZATCA::UBL::Invoice::PAYMENT_MEANS[:bank_card] # => "48"
invoice_type = ZATCA::UBL::Invoice::TYPES[:invoice] # => "388"

invoice_counter_value = "10"
previous_invoice_hash = "NWZlY2ViNjZmZmM4NmYzOGQ5NTI3ODZjNmQ2OTZjNzljMmRiYzIzOWRkNGU5MWI0NjcyOWQ3M2EyN2ZiNTdlOQ=="
currency_code = "SAR"
vat_registration_number = "311111111101113"

#===============================================================================
# 2. Setup the Invoice's Nested Properties
#===============================================================================

# Create the supplier party (the party issuing the invoice, e.g. the seller)
accounting_supplier_party = ZATCA::UBL::CommonAggregateComponents::Party.new(
  party_identification: ZATCA::UBL::CommonAggregateComponents::PartyIdentification.new(
    id: "324223432432432"
  ),
  postal_address: ZATCA::UBL::CommonAggregateComponents::PostalAddress.new(
    street_name: "الامير سلطان",
    additional_street_name: nil,
    building_number: "3242",
    plot_identification: "4323",
    city_subdivision_name: "32423423",
    city_name: "الرياض | Riyadh",
    postal_zone: "32432",
    country_subentity: nil,
    country_identification_code: "SA"
  ),

  party_tax_scheme: ZATCA::UBL::CommonAggregateComponents::PartyTaxScheme.new(
    company_id: vat_registration_number
  ),

  party_legal_entity: ZATCA::UBL::CommonAggregateComponents::PartyLegalEntity.new(
    registration_name: "Acme Widgets LTD"
  )
)

# Create the customer party (the party receiving the invoice, e.g. the buyer)
accounting_customer_party = ZATCA::UBL::CommonAggregateComponents::Party.new(
  # party_identification: ZATCA::UBL::CommonAggregateComponents::PartyIdentification.new(
  #   id: "2345",
  #   scheme_id: "NAT"
  # ),
  party_identification: nil,
  postal_address: ZATCA::UBL::CommonAggregateComponents::PostalAddress.new(
    street_name: nil,
    additional_street_name: nil,
    building_number: nil,
    plot_identification: nil,
    city_subdivision_name: "32423423",
    city_name: nil,
    postal_zone: nil,
    country_subentity: nil,
    country_identification_code: "SA"
  ),

  party_tax_scheme: ZATCA::UBL::CommonAggregateComponents::PartyTaxScheme.new,
  party_legal_entity: nil
)

# Create the (optional) delivery object (detailing the delivery date and time)
# delivery = ZATCA::UBL::CommonAggregateComponents::Delivery.new(
#   actual_delivery_date: "2022-03-13",
#   latest_delivery_date: "2022-03-15"
# )

delivery = nil

# Create the allowance charges (e.g. discounts) for the invoice
allowance_charges = [
  ZATCA::UBL::CommonAggregateComponents::AllowanceCharge.new(
    charge_indicator: false,
    amount: "0.00",
    allowance_charge_reason: "discount",
    currency_id: "SAR",
    tax_categories: [
      # Yes, ZATCA's official valid sample duplicates these, not sure why
      ZATCA::UBL::CommonAggregateComponents::TaxCategory.new(
        tax_percent: "15"
      ),
      ZATCA::UBL::CommonAggregateComponents::TaxCategory.new(
        tax_percent: "15"
      )
    ]
  )
]

# Create the tax totals for the invoice
# ZATCA's official valid sample has two of these, not sure why
tax_totals = [
  ZATCA::UBL::CommonAggregateComponents::TaxTotal.new(
    tax_amount: "30.15"
  ),
  ZATCA::UBL::CommonAggregateComponents::TaxTotal.new(
    tax_amount: "30.15",
    tax_subtotal_amount: "30.15",
    taxable_amount: "201.00",
    tax_category: ZATCA::UBL::CommonAggregateComponents::TaxCategory.new(
      tax_percent: "15.00"
    )
  )
]

# Create the legal monetary total for the invoice
legal_monetary_total = ZATCA::UBL::CommonAggregateComponents::LegalMonetaryTotal.new(
  line_extension_amount: "201.00",
  tax_exclusive_amount: "201.00",
  tax_inclusive_amount: "231.15",
  allowance_total_amount: "0.00",
  prepaid_amount: "0.00",
  payable_amount: "231.15"
)

# Create the invoice lines for the invoice (the list of items that were sold)
invoice_lines = [
  # Book
  ZATCA::UBL::CommonAggregateComponents::InvoiceLine.new(
    invoiced_quantity: "33.000000",
    invoiced_quantity_unit_code: "PCE",
    line_extension_amount: "99.00",
    tax_total: ZATCA::UBL::CommonAggregateComponents::TaxTotal.new(
      tax_amount: "14.85",
      rounding_amount: "113.85"
    ),
    item: ZATCA::UBL::CommonAggregateComponents::Item.new(
      name: "كتاب"
    ),
    price: ZATCA::UBL::CommonAggregateComponents::Price.new(
      price_amount: "3.00",
      allowance_charge: ZATCA::UBL::CommonAggregateComponents::AllowanceCharge.new(
        charge_indicator: false,
        allowance_charge_reason: "discount",
        amount: "0.00",
        add_tax_category: false,

        # ZATCA's samples can sometimes have a nested tax scheme with an ID
        # and sometimes they omit it. Setting this boolean controls whether it is
        # present or not
        add_id: false
      )
    )
  ),

  # Pen
  ZATCA::UBL::CommonAggregateComponents::InvoiceLine.new(
    invoiced_quantity: "3.000000",
    invoiced_quantity_unit_code: "PCE",
    line_extension_amount: "102.00",
    tax_total: ZATCA::UBL::CommonAggregateComponents::TaxTotal.new(
      tax_amount: "15.30",
      rounding_amount: "117.30"
    ),
    item: ZATCA::UBL::CommonAggregateComponents::Item.new(
      name: "قلم"
    ),
    price: ZATCA::UBL::CommonAggregateComponents::Price.new(
      price_amount: "34.00",
      allowance_charge: ZATCA::UBL::CommonAggregateComponents::AllowanceCharge.new(
        charge_indicator: false,
        allowance_charge_reason: "discount",
        amount: "0.00",
        add_tax_category: false,
        add_id: false
      )
    )
  )
]

#===============================================================================
# 3. Construct the Invoice
#===============================================================================
# Construct the invoice using all of the above
invoice = ZATCA::UBL::Invoice.new(
  add_ids_to_allowance_charges: false,
  id: invoice_id,
  uuid: invoice_uuid,
  issue_date: issue_date,
  issue_time: issue_time,
  subtype: invoice_subtype,
  type: invoice_type,
  invoice_counter_value: invoice_counter_value,
  previous_invoice_hash: previous_invoice_hash,
  accounting_supplier_party: accounting_supplier_party,
  accounting_customer_party: accounting_customer_party,
  delivery: delivery,
  payment_means_code: payment_means_code,
  allowance_charges: allowance_charges,
  tax_totals: tax_totals,
  legal_monetary_total: legal_monetary_total,
  invoice_lines: invoice_lines,
  currency_code: currency_code,
  note: note,
  note_language_id: note_language_id
)

# NOTE: For credit or debit notes you must also fill out
# invoice.instruction_note with an explanation on why
# this note was issued.

#===============================================================================
# 4. Sign the Invoice (ONLY FOR SIMPLIFIED INVOICES)
#===============================================================================

# Must not be encoded in Base64 and must have header blocks
private_key_path = "ec-secp256k1-priv-key.pem"

# Must be in pem format with header blocks
certificate_path = "cert.pem"

# You can omit this if you don't need to customize it,
# the default value is the same as you see here.
signing_time = Time.now.utc.strftime("#{invoice.issue_date}T#{invoice.issue_time}")

# Sign the invoice (this adds a couple of new elements to the invoice)

invoice.sign(
  private_key_path: private_key_path,
  certificate_path: certificate_path,
  signing_time: signing_time,
  decode_private_key_from_base64: false # Optional
)

# See the section under this code block if you need more
# control over the signing process.


#===============================================================================
# 5. Create the QR Code (ONLY FOR SIMPLIFIED INVOICES)
#===============================================================================

# Setup values used in tags
invoice_hash = invoice.generate_hash
invoice_timestamp = "#{invoice.issue_date}T#{invoice.issue_time}"

# Create the tags
tags = ZATCA::Tags.new({
                         seller_name: ENV.fetch('SELLER_NAME', 'Maximum Speed Tech Supply LTD),
                         vat_registration_number: ENV.fetch('INAC_VAT_REGISTRATION_NUMBER', '399999999900003'),
                         timestamp: invoice_timestamp,
                         vat_total: "30.15",
                         invoice_total: "231.15",
                         xml_invoice_hash: invoice_hash,

                         # These 3 properties on the invoice assume you have signed it before.
                         # They are nil unless signed.
                         ecdsa_signature: invoice.signed_hash,
                         ecdsa_public_key: invoice.public_key_bytes,
                         ecdsa_stamp_signature: invoice.certificate_signature
                       })

# Turn the tags to TLV and then encode that TLV to base64
invoice.qr_code = tags.to_base64


File.write("signed_invoice.xml", invoice.to_xml)

puts "finish"

**Comment by asifmeer** [comment]
i copied the code from here: https://github.com/mrsool/zatca/wiki/Constructing-a-Simplified-Invoice
updated certificates path and used seller name and var_registration from certificates.
in last line i added this save xml response in file
File.write("signed_invoice.xml", invoice.to_xml)

here is command to validate file.
fatoora -validate -invoice signed_invoice.xml


gem code is generating wrong values, or might be i am doing somthing wrong.

**Reply by obahareth** [reply]
@asifmeer It looks like you wrote the QR code to a file called `signed_invoice.xml` then tried to validate the QR code as an XML file.

You have two options here:
1. Use ZATCA's APIs to do the validation, check the documentation on the compliance endpoint
2. Save the invoice (instead of the QR Code) as an XML file.


For the second option, after signing the invoice you can simply call:
`invoice.generate_xml`

**Reply by asifmeer** [reply]
Dear @obahareth ,

I wanted to take a moment to express my sincere gratitude for the support you provided with the Zatca gem. Your assistance was instrumental in resolving the issues I encountered, and I truly appreciate the time and effort you invested in helping me.

The Zatca gem has proven to be a powerful tool in my Ruby on Rails project, and your guidance has greatly enhanced my experience. I am excited to share my positive experience with this gem through upcoming blog posts, which I hope will contribute to its continued success and benefit other developers.

Thank you once again for your exceptional support and dedication.

Best regards,
Asif Qasim
